### PR TITLE
Scalar refactor

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Remove `Int`, `Bool`, `Double`, and `ComplexDouble`. Use a single non-template class `Scalar` for all scalars. This simplifies many things up, and we are 900+ lines of code less.

### CODE REVIEW

File: csrc/ir/base_nodes.cpp [hunk 1]
```diff
@@ -330,22 +322,34 @@ bool Val::evaluateBool() {
 }
 
 std::optional<int64_t> Val::getInt() const {
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.cpp:324]: In a followup PR, I will replace this with a `getValue`

**Comment by zasdfgbnm** [comment]
!build

File: tools/examples/repro.py
Comment (zasdfgbnm) [inline tools/examples/repro.py:1]: This file is unformatted otherwise the script will have trouble parsing

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/builder.h [hunk 2]
```diff
@@ -56,42 +56,42 @@ class TORCH_CUDA_CU_API IrBuilder {
   static T* clone(const T* src, IrCloner* ir_cloner);
 
   // Unary operations
-  static Val* negExpr(Val* val);
-  static Val* notExpr(Val* val);
-  static Val* absExpr(Val* val);
-  static Val* setExpr(Val* val);
-  static Val* setExprNamedScalar(const std::string& name, Val* val);
-  static Val* addressExprNamedScalar(const std::string& name, Val* val);
+  static Scalar* negExpr(Val* val);
```
Comment (naoyam) [inline csrc/ir/builder.h:59]: Wonder if the input parameter should also be of type `Scalar`

File: csrc/ir/nodes.cpp [hunk 3]
```diff
@@ -32,75 +32,7 @@
 
 namespace nvfuser {
 
-namespace {
```
Comment (naoyam) [inline csrc/ir/nodes.cpp:35]: Nice cleanup!

**Review by naoyam — COMMENTED** [review]
LGTM. I'll let @jjsjann123 and @rdspring1 to check the changes for the Python interface and the serialization.

**Comment by zasdfgbnm** [comment]
!build

**Review by jjsjann123 — COMMENTED** [review]
LGTM on python API side. Would @rdspring1 mind taking a look at serde stuff?

File: csrc/scalar_value.h [hunk 4]
```diff
@@ -50,18 +49,19 @@ struct Struct {
...
-  size_t size_;
+  int64_t size_;
 
  public:
   template <typename T>
   Pointer(T* ptr) : ptr_(reinterpret_cast<std::byte*>(ptr)), size_(sizeof(T)) {}
 
-  Pointer(void* ptr, DataType dtype)
-      : ptr_(reinterpret_cast<std::byte*>(ptr)), size_(dataTypeSize(dtype)) {}
+  inline Pointer(void* ptr, DataType dtype);
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/scalar_value.h]: This is a cyclic dependency issue: `DataType` is defined in `type.h`, but this file is included by `type.h`, so we have to put the definition into `type.h`.

File: csrc/type.h [hunk 5]
```diff
@@ -829,4 +835,32 @@ const char* const kMagicZeroName = "nvfuser_zero";
...
+          value = ScalarValue(static_cast<T>(value));
+        }
+      }
+      // TODO: support arrays and pointers
+    });
+  }
+  TORCH_CHECK(
+      isCompatibleDataType(nvfuser::getDataType(value), dtype),
+      "Scalar value is not compatible with the given data type.");
+  return value;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/type.h:865]: The cast actually happens above:
```C++
value = PolymorphicValue(static_cast<T>(value));
```

File: csrc/serde/utils.cpp [hunk 6]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    const nvfuser::PolymorphicValue& v,
+    serde::DataType t) {
+  ScalarBuilder builder_(builder);
+  builder_.add_dtype(t);
+  if (v.is<std::monostate>()) {
+    builder_.add_has_value(false);
+    return builder_.Finish();
+  } else if (v.is<double>()) {
+    builder_.add_has_value(true);
+    builder_.add_value_type(serde::DataType::DataType_Double);
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:136]:
```suggestion
    builder_.add_value_type(serde::DataType_Double);
```

File: csrc/serde/utils.cpp [hunk 7]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    builder_.add_double_value(v.as<double>());
+    return builder_.Finish();
+  } else if (v.is<int64_t>()) {
+    builder_.add_has_value(true);
+    builder_.add_value_type(serde::DataType::DataType_Int);
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:141]:
```suggestion
    builder_.add_value_type(serde::DataType_Int);
```

File: csrc/serde/utils.cpp [hunk 8]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    builder_.add_long_value(v.as<int64_t>());
+    return builder_.Finish();
+  } else if (v.is<bool>()) {
+    builder_.add_has_value(true);
+    builder_.add_value_type(serde::DataType::DataType_Bool);
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:146]:
```suggestion
    builder_.add_value_type(serde::DataType_Bool);
```

File: csrc/serde/utils.cpp [hunk 9]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    builder_.add_bool_value(v.as<bool>());
+    return builder_.Finish();
+  } else if (v.is<std::complex<double>>()) {
+    builder_.add_has_value(true);
+    auto c = v.as<std::complex<double>>();
+    builder_.add_value_type(serde::DataType::DataType_ComplexDouble);
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:152]:
```suggestion
    builder_.add_value_type(serde::DataType_ComplexDouble);
```

File: csrc/serde/utils.cpp [hunk 10]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    return builder_.Finish();
+  }
+  TORCH_INTERNAL_ASSERT(
+      false, "Unable to convert ", v.type().name(), " to serde::Scalar.");
+}
+
+PolymorphicValue parsePolymorphicValue(const serde::Scalar* c) {
+  if (!c->has_value()) {
+    return {};
+  } else if (c->value_type() == DataType::DataType_Double) {
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:164]:
```suggestion
  } else if (c->value_type() == serde::DataType_Double) {
```

File: csrc/serde/utils.cpp [hunk 11]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    return PolymorphicValue(c->double_value());
+  } else if (c->value_type() == DataType::DataType_Int) {
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:166]:
```suggestion
  } else if (c->value_type() == serde::DataType_Int) {
```

File: csrc/serde/utils.cpp [hunk 12]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    return PolymorphicValue(c->long_value());
+  } else if (c->value_type() == DataType::DataType_Bool) {
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:168]:
```suggestion
  } else if (c->value_type() == serde::DataType_Bool) {
```

File: csrc/serde/utils.cpp [hunk 13]
```diff
@@ -122,6 +122,59 @@ PrimDataType mapToNvfuserDtype(serde::DataType t) {
...
+    return PolymorphicValue(c->bool_value());
+  } else if (c->value_type() == DataType::DataType_ComplexDouble) {
```
Comment (rdspring1) [inline csrc/serde/utils.cpp:170]:
```suggestion
  } else if (c->value_type() == serde::DataType_ComplexDouble) {
```

File: csrc/python_frontend/fusion_record.h [hunk 14]
```diff
@@ -2044,106 +2028,23 @@ struct ScalarRecord : RecordFunctor {
 
   std::pair<serde::RecordData, flatbuffers::Offset<void>> recordData(
       flatbuffers::FlatBufferBuilder& builder) const final {
-    return valueRecordData(builder, value_);
-  };
+    return {
+        serde::RecordData_Scalar,
+        serde::serializeScalar(builder, value_, serde::mapToSerdeDtype(dtype_))
```
Comment (rdspring1) [inline csrc/python_frontend/fusion_record.h:2033]: Nitpick: I think it is slightly cleaner to pass `dtype_` as-is and then convert to `serde dtype` in `serializeScalar`.

**Comment by rdspring1** [comment]
LGTM

**Review by rdspring1 — APPROVED** [review]
LGTM

File: csrc/python_frontend/fusion_record.h (hunk 14 above)
Comment (zasdfgbnm) [reply csrc/python_frontend/fusion_record.h]: fixed

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**